tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc

```diff
@@ -0,0 +1,2918 @@
+
+  if (input_is_qtype != output_is_qtype) {
+    op->emitOpError(
+        "ConvertTFLRelu6Op: input/output tensor should "
+        "be all quantized or all native");
+    return failure();
+  }
+
+  Value output;
+  if (output_is_qtype) {
+    auto rescale_type =
+        RankedTensorType::get(output_type.getShape(), rewriter.getI32Type());
+    auto input_qtype = input_type.getElementType()
+                           .dyn_cast<mlir::quant::UniformQuantizedType>();
+    auto output_qtype = output_type.getElementType()
+                            .dyn_cast<mlir::quant::UniformQuantizedType>();
+    int64_t rescaled_6 = (int64_t)std::round(6.0f / input_qtype.getScale()) +
+                         input_qtype.getZeroPoint();
+
+    auto op1_rescale_in = buildRescaleToInt32(rewriter, op, tfl_relu6_op.x(),
+                                              1.0f, input_qtype.getZeroPoint());
+    auto op2_relun_op1 = rewriter.create<tosa::ReluNOp>(
+        op->getLoc(), rescale_type, op1_rescale_in,
+        rewriter.getI64IntegerAttr(rescaled_6), rewriter.getF32FloatAttr(0.0f));
+    auto op3_rescale_op2 = buildRescaleFromInt32(
+        rewriter, op, output_type, op2_relun_op1.getResult(), 1.0f,
+        output_qtype.getZeroPoint());
+
+    output = op3_rescale_op2;
+  } else {
+    auto op1_relun_in = rewriter.create<tosa::ReluNOp>(
+        op->getLoc(), output_type, tfl_relu6_op.x(),
+        rewriter.getI64IntegerAttr(0), rewriter.getF32FloatAttr(6.0f));
+
+    output = op1_relun_in.getResult();
+  }
+
+  rewriter.replaceOp(op, {output});
+  return success();
+}
+
+LogicalResult ConvertTFLEqualOp::matchAndRewrite(
+    Operation* op, PatternRewriter& rewriter) const {
+  auto tfl_equal_op = cast<TFL::EqualOp>(op);
+
+  auto input_x_type = tfl_equal_op.x().getType().dyn_cast<RankedTensorType>();
+  auto input_y_type = tfl_equal_op.y().getType().dyn_cast<RankedTensorType>();
+  auto output_type =
+      tfl_equal_op.getResult().getType().dyn_cast<RankedTensorType>();
+  // Not a ranked tensor output
+  if (!input_x_type || !input_y_type || !output_type) return failure();
+
+  bool input_x_is_qtype =
+      input_x_type.getElementType().isa<mlir::quant::UniformQuantizedType>();
+  bool input_y_is_qtype =
+      input_y_type.getElementType().isa<mlir::quant::UniformQuantizedType>();
+  bool output_is_qtype =
+      output_type.getElementType().isa<mlir::quant::UniformQuantizedType>();
+
+  if (input_x_is_qtype != output_is_qtype ||
+      input_y_is_qtype != output_is_qtype) {
+    op->emitOpError(
+        "ConvertTFLEqualOp: input/output tensor should "
+        "be all quantized or all native");
+    return failure();
+  }
+
+  Value output;
+  if (output_is_qtype) {
+    auto rescale_type =
+        RankedTensorType::get(output_type.getShape(), rewriter.getI32Type());
+    auto input_x_qtype = input_x_type.getElementType()
+                             .dyn_cast<mlir::quant::UniformQuantizedType>();
```

Comment:
Hoist out and condition on.

Reply:
Same suggested alternative above for use here and in following instances.